### ISSUE TITLE
Metrics in staggered grids

### ISSUE THREAD

**Issue by johnomotani** [boutproject/BOUT-dev#64]
In difops.cxx, the CELL_CENTRE metrics are used regardless of the CELL_LOC of the derivative. This will introduce an error into the results when using staggered grids if the metric varies (in the direction of the stagger).

The solution, as it seems to me, would be to keep CELL_XLOW and CELL_YLOW versions of the metric in Mesh (either provided by the grid generator or calculated by interpolation), which can then be used as needed in difops.cxx.

An alternative, if we want to make staggered grids really first-class citizens in BOUT++, might be to associate fields at different locations with the Field*D object and introduce a method like

```
Field*D & operator()(CELL_LOC loc);
```

which would return the Field at `loc` (then `Field*D::setLocation` might just set the 'default' location). We could either interpolate from the default location when this operator is called (if the field has changed since it was last called), or add some method to the Field that the user must call to calculate the non-default versions. The differential operators might then change very simply, e.g.

```
result = DDY(var, outloc, method)/sqrt(mesh->g_22);
```

becomes

```
result = DDY(var, outloc, method)/sqrt( mesh->g_22(outloc) );
```

This seems like it might be an elegant solution, but would probably require quite a bit of work to implement...

**Comment by dschwoerer** [comment]
Don't non-uniform meshes produce anyway errors, as the stencils (to my understanding) all assume a uniform mesh?
If the mesh spacing varies that strongly, I am not sure this is your most important problem ...

**Comment by dschwoerer** [comment]
Partially fixed by the [newfield-branch](https://github.com/boutproject/BOUT-dev/tree/newfield)

**Comment by bendudson** [comment]
Is this now fixed with the new coordinates at different locations?

**Comment by johnomotani** [comment]
I think it should be fixed... I'm working on an MMS test for differential operators to verify the staggered versions.

**Comment by johnomotani** [comment]
Fixed in #1198/#1271
